statscache_plugins/volume/by_category.py

```diff
@@ -14,28 +13,31 @@ class PluginMixin(VolumePluginMixin):
     the bus for each category.
     """
 
-    def handle(self, session, messages):
-        volumes = collections.defaultdict(int)
-        for msg in messages:
-            msg_timestamp = datetime.datetime.fromtimestamp(msg['timestamp'])
-            volumes[(msg['topic'].split('.')[3],
-                     self.frequency.next(now=msg_timestamp))] += 1
+    def process(self, message):
+        timestamp = self.schedule.next(
+            now=datetime.datetime.fromtimestamp(message['timestamp'])
+        )
+        category = message['topic'].split('.')[3]
+        self._volumes[(category, timestamp)] += 1
 
-        for key, volume in volumes.items():
+    def update(self, session):
```

Comment:
Very true, these methods are boilerplate code to a large extent (or at least, the volume ones are). Yep, your understanding of that is correct; plugins need only know how to process a message and update the model. I'll play around with this and push a change to consolidate the common logic into the ``VolumePluginMixin`` class, most likely.